src/gdft.c

```diff
@@ -267,168 +260,172 @@ gdMutexDeclare (gdFontCacheMutex);
+			*chPtr = (uint32_t)byte;
 			return 1;
 		}
+	}
+	*chPtr = (uint32_t)UnicodeTbl[ku - 1][ten - 1];
+	return 2;
+}
 
-		*chPtr = (Tcl_UniChar) UnicodeTbl[ku - 1][ten - 1];
-		return 2;
-	} else
-#endif /* JISX0208 */
-		if (byte < 0xC0) {
-			/*
-			 * Handles properly formed UTF-8 characters between
-			 * 0x01 and 0x7F.  Also treats \0 and naked trail
-			 * bytes 0x80 to 0xBF as valid characters representing
-			 * themselves.
-			 */
-
-			*chPtr = (Tcl_UniChar) byte;
-			return 1;
-		} else if (byte < 0xE0) {
-			if ((str[1] & 0xC0) == 0x80) {
-				/*
-				 * Two-byte-character lead-byte followed
-				 * by a trail-byte.
-				 */
-
-				*chPtr = (Tcl_UniChar) (((byte & 0x1F) << 6) | (str[1] & 0x3F));
-				return 2;
-			}
-			/*
-			 * A two-byte-character lead-byte not followed by trail-byte
-			 * represents itself.
-			 */
-
-			*chPtr = (Tcl_UniChar) byte;
-			return 1;
-		} else if (byte < 0xF0) {
-			if (((str[1] & 0xC0) == 0x80) && ((str[2] & 0xC0) == 0x80)) {
-				/*
-				 * Three-byte-character lead byte followed by
-				 * two trail bytes.
-				 */
+#else
 
-				*chPtr = (Tcl_UniChar) (((byte & 0x0F) << 12)
-				                        | ((str[1] & 0x3F) << 6) | (str[2] & 0x3F));
-				return 3;
-			}
-			/*
-			 * A three-byte-character lead-byte not followed by
-			 * two trail-bytes represents itself.
-			 */
+static int gd_UTF8_To_Unicode(const char *str, uint32_t *chPtr) {
+	// reads the contents pointed to by str and outputs a unicode codepage into object pointed to by chPtr
+	// return is the bytes to offset for the first char of a subsequent read
+	uint32_t b0 = *(uint8_t *)str;
+	uint32_t b1, b2, b3;
 
-			*chPtr = (Tcl_UniChar) byte;
-			return 1;
+	if (b0 < 0xC0) {
+		*chPtr = b0;
+		return 1;
+	}
+	if (b0 < 0xE0) {
+		b1 = (uint8_t)str[1];
+		if ((b1 & 0xC0) == 0x80) {
+			*chPtr = ((b0 & 0x1F) << 6) | (b1 & 0x3F);
+			return 2;
 		}
-#if TCL_UTF_MAX > 3
-		else {
-			int ch, total, trail;
-
-			total = totalBytes[byte];
-			trail = total - 1;
-			if (trail > 0) {
-				ch = byte & (0x3F >> trail);
-				do {
-					str++;
-					if ((*str & 0xC0) != 0x80) {
-						*chPtr = byte;
-						return 1;
-					}
-					ch <<= 6;
-					ch |= (*str & 0x3F);
-					trail--;
-				} while (trail > 0);
-				*chPtr = ch;
-				return total;
-			}
+		*chPtr = b0;
+		return 1;
+	}
+	if (b0 < 0xF0) {
+		b1 = (uint8_t)str[1];
+		b2 = (uint8_t)str[2];
+		if (((b1 & 0xC0) == 0x80) && ((b2 & 0xC0) == 0x80)) {
+			*chPtr = ((b0 & 0x0F) << 12) | ((b1 & 0x3F) << 6) | (b2 & 0x3F);
+			return 3;
 		}
-#endif
-
-	*chPtr = (Tcl_UniChar) byte;
+		*chPtr = b0;
+		return 1;
+	}
+	if (b0 < 0xF8) {
+		b1 = (uint8_t)str[1];
+		b2 = (uint8_t)str[2];
+		b3 = (uint8_t)str[3];
+		if (((b1 & 0xC0) == 0x80) && ((b2 & 0xC0) == 0x80) && ((b3 & 0xC0) == 0x80)) {
+			*chPtr = ((b0 & 0x7) << 18) | ((b1 & 0x3F) << 12) | ((b2 & 0x3F) << 6) | (b3 & 0x3F);
+			return 4;
+		}
+		*chPtr = b0;
+		return 1;
+	}
+	printf("WARNING: utf-8 above 21-bit unicode range\n");
```

Comment:
you want to include `gd_errors.h` and use `gd_error()` here instead

Reply:
Do you want just a warning or a fatal error ?
Previous behavior would continue untill the entire loop finished.
So any bad lead byte would and does make any following bytes nonsense.
imo a warning seems like enough, the user is likely to notice.

Reply:
as a library, we should very very rarely be fatal.  it makes users a lot harder to be reliable.

a warning in this specific case sounds fine to me.